HIP/hip-1028.md

```diff
@@ -0,0 +1,157 @@
+---
+hip: 1028
+title: Metadata management via SmartContracts
+author: Mustafa Uzun <mustafa.uzun@limechain.tech>, Luke Lee <luke.lee@swirldslabs.com>
+working-group: Valentin Valkanov <valentin.valkanov@limechain.tech>, David Bakin <david.bakin@swirldsbals.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2024-08-15
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/pull/1028
+requires: 646, 657, 765
+updated: 2024-08-19
+requested-by: NA
+---
+
+## Abstract
+
+HIP-646/657/765 introduces a new metadata field for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key  
+for updating metadata. However, these features are not supported by smart contracts.  
+This proposal aims to enhance Hedera Smart Contract Service (HSCS) by exposing HIP-646/657/765 features.
+
+## Motivation
+
+A user may need to create, update and get info for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key through contract.  
+The current version of Hedera Smart Contract Service (HSCS) does not cover these use cases.
+
+## Rationale
+
+The HSCS already has the functionality to create and update tokens, so we would want to extend these.  
+
+However, we will add new functions for getting info for tokens to not break the older version of ``HederaTokenService.sol``  
+
+New function  for TokenUpdateNfts HAPI operation will be added  
+
+## User stories
+
+1. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with metadata.
+2. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata.
+3. As an EOA I would like to initiate a smart contract transaction that updates a fungible token with metadata.
+4. As an EOA I would like to initiate a smart contract transaction that updates a non-fungible token collection with metadata.
+5. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with a metadata key.
+6. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata key.
+7. As an EOA I would like to initiate a smart contract transaction that updates a fungible token with metadata key.
+8. As an EOA I would like to initiate a smart contract transaction that updates a non-fungible token collection with metadata key.
+9. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with metadata and metadata key.
+10. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata and metadata key.
```

Comment:
I know it is the practice in our existing HIPs to just spell out all these variations in a linear list like this, but especially in this particular case I think it could be done differently.  And better:  Right now this list 1..10 looks to me like "twisty passages all alike".  When it is really a 2x2x2 table with an extra 2 that don't quite fit (9&10).

Would it be too different to start with something like this table and then make it better?  Would that be easier to understand all the variations?  It could even be more compact by coalescing Create+Update columns (use `C` and `U` as the column markers) and Fungible+Non-Fungible columns (use `F` and `N` as the markers?).  Or some other kind of table?

|Create|Update|Fungible|Non-Fungible||Metadata?|Metadata Key?|
|----|----|----|----|----|----|----|
| ◉ || ◉ ||| ✅ ||
| ◉ || ◉ |||| ✅ |
| ◉ || ◉ ||| ✅ | ✅ |
| ◉ ||| ◉ || ✅ ||
| ◉ ||| ◉ ||| ✅ |
| ◉ ||| ◉ || ✅ | ✅ |
|| ◉ | ◉ ||| ✅ ||
|| ◉ | ◉ |||| ✅ |
|| ◉ || ◉ || ✅ ||
|| ◉ || ◉ ||| ✅ |

Reply:
How about including both?  I like the table in the context of the linear list of user stories.